staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go

```diff
@@ -534,14 +534,14 @@ func (w *watchCache) GetAllEventsSinceThreadUnsafe(resourceVersion uint64) ([]*w
 	size := w.endIndex - w.startIndex
 	var oldest uint64
 	switch {
+	case w.listResourceVersion > 0 && w.startIndex == 0:
+		// If no event wasn't removed from the buffer since last relist, the oldest watch
+		// event we can deliver is one greater than the resource version of the list.
+		oldest = w.listResourceVersion + 1
 	case size >= w.capacity:
```

Comment:
How about changing case `size >= w.capacity` and `size > 0` into 'size > 0 && size <= w.capacity'?

Reply:
BTW, can listResourceVersion was set as `w.cache[w.startIndex%w.capacity].ResourceVersion` when its size changed, which one is better?

Reply:
> Rather than change GetAllEventsSinceThreadUnsafe, listResourceVersion can be set as w.cache[w.startIndex%w.capacity].ResourceVersion when its size changed, which one is better?

I consciously didn't go this way. I want to leave the semantic of `listResourceVersion` really the RV of last list request. Not something random.

> One stupid question, does resourceVersion face similar issue?

I don't understand this question.

Reply:
> How about changing case size >= w.capacity and size > 0 into size > 0 && size <= w.capacity?

Updated and updated the comment.

Reply:
Thanks for digging into and fixing it.